addons/package_hierarchy/models/package_links.py

```diff
@@ -0,0 +1,185 @@
+        move_lines |= move_lines.mapped("move_id").mapped("move_line_ids")
+        links = move_lines.mapped("u_result_package_link_ids")
+        # If some links are not associated with movelines, we still want to check them:
+        links |= self
+        links._validate_links()
+
+    def _validate_links(self):
+        """Validate package links to ensure that no constraints are broken."""
+        max_package_depth = self.env.user.get_user_warehouse().u_max_package_depth
+
+        # Sanitize links, check for repeated children as we should not have any
+        # Only exception may be moving a package from one package into another package. This will
+        # create a 'un-link' (no-parent) and a 'link' (with parent).
+        children_count = {}
+        for link in self:
+            children_count[link.child_id] = children_count.get(link.child_id, 0) + 1
+        for child_id, count in children_count.items():
+            if count == 2:
+                repeated_links = self.filtered(lambda x: x.child_id == child_id)
+                parent_ids = [link.parent_id.id for link in repeated_links]
+                if parent_ids.count(False) != 1:
+                    raise ValidationError(_("Links are proposing to move package "
+                                            "to multiple locations."))
+            elif count > 2:
+                raise ValidationError(_("Links are proposing to move package "
+                                        "to multiple locations."))
+
+        # Make sure all links have child_id
+        if not all(link.child_id for link in self):
+            raise ValidationError(_("Links should always have a child."))
+
+        # Create chains from the links
+        chains = self._return_chains()
+        # Create a dict of links for efficient searching for unlinks later
+        unlinks_from_child = {link.child_id: link for link in self if not link.parent_id}
+
+        # Now need to check these chains against the current reality
+        # NOTE: Currently this is done for a single chain at a time, if we wanted to be
+        # thorough we should check for multiple chains within the same tree at once. This
+        # would only occur for very complex moves so the computational effort may not be
+        # worth it for a 'normal' system.
+        for chain in chains:
+            # Check for self-ancestors
+            current_ancestors = chain[-1]._return_ancestors()
+            if any(True for node in chain if node in current_ancestors):
+                raise ValidationError(_("Proposed link(s) would result in a package loop"))
+
+            # Check the depth of the proposed tree
+            # Check if there is a relevant unlink
+            if unlinks_from_child.get(chain[-1], False):
+                length_above = 0
+            else:
+                length_above = chain[-1]._return_num_ancestors()
+            total_length = length_above + len(chain) + chain[0].depth - 1
```

Comment:
What happens when we have irregular trees? When one of the nodes of the chain has a higher depth than: `(its_distance_from_chain[0]) + chain[0].depth`. Do we want to add an comment that we are assuming regular trees?

Reply:
Good spot, I'll have a think on this and will at minimum add a comment

Reply:
I've changed the function to account for this and added a new test